wallet-new/test/unit/Test/Spec/Addresses.hs

```diff
@@ -348,6 +395,137 @@ spec = describe "Addresses" $ do
+                        case res of
+                            Right wr | null (wrData wr) -> pure ()
+                            _ -> fail ("Got " ++ show res)
+
+            prop "it yields the correct number of results" $ withMaxSuccess 20 $ do
+                monadicIO $
+                    withAddressesFixtures 3 4 $ \_ layer _ _ -> do
+                        let pp = PaginationParams (Page 1) (PerPage 40)
+                        res <- runExceptT $ runHandler' $ do
+                            Handlers.listAddresses layer (RequestParams pp)
+                        case res of
+                            Right wr -> do
+                                -- this takes into account that there is an initial account
+                                -- and each account has an initial address (but not the initial
+                                -- account thus the -1)
+                                length (wrData wr) `shouldBe` expectedNumber 3 4
+                            _        -> fail ("Got " ++ show res)
+
+            prop "is deterministic" $ withMaxSuccess 20 $ do
+                monadicIO $
+                    withAddressesFixtures 3 8 $ \_ layer _ _ -> do
+                        let (expectedTotal :: Int) = expectedNumber 3 8
+                        let pp = PaginationParams (Page 1) (PerPage 40)
+                        let pp1 = PaginationParams (Page 1) (PerPage (quot expectedTotal 3 + 1))
+                        let pp2 = PaginationParams (Page 2) (PerPage (quot expectedTotal 3 + 1))
+                        let pp3 = PaginationParams (Page 3) (PerPage (quot expectedTotal 3 + 1))
+                        res <- runExceptT $ runHandler' $ do
+                            Handlers.listAddresses layer (RequestParams pp)
+                        res' <- runExceptT $ runHandler' $ do
+                            Handlers.listAddresses layer (RequestParams pp)
+                        res1 <- runExceptT $ runHandler' $ do
+                            Handlers.listAddresses layer (RequestParams pp1)
+                        res1' <- runExceptT $ runHandler' $ do
+                            Handlers.listAddresses layer (RequestParams pp1)
+                        res2 <- runExceptT $ runHandler' $ do
+                            Handlers.listAddresses layer (RequestParams pp2)
+                        res2' <- runExceptT $ runHandler' $ do
+                            Handlers.listAddresses layer (RequestParams pp2)
+                        res3 <- runExceptT $ runHandler' $ do
+                            Handlers.listAddresses layer (RequestParams pp3)
+                        res3' <- runExceptT $ runHandler' $ do
+                            Handlers.listAddresses layer (RequestParams pp3)
+                        res `shouldBe` res'
+                        res1 `shouldBe` res1'
+                        res2 `shouldBe` res2'
+                        res3 `shouldBe` res3'
```

Comment:
Idea: considering is very error prone to go for this style of `foo, foo', bar, bar'`, why don't you write something like this?

``` haskell
                        res <- runExceptT $ runHandler' $ do
                            r1 <- Handlers.listAddresses layer (RequestParams pp)
                            r2 <- Handlers.listAddresses layer (RequestParams pp)
                            pure (r1 === r2)
                        res1 <- runExceptT $ runHandler' $ do
                            r1 <- Handlers.listAddresses layer (RequestParams pp1)
                            r2 <- Handlers.listAddresses layer (RequestParams pp1)
                            pure (r1 === r2)
                       assert $ conjoin [res, res1]
```

Even better, I think you can simplify this even further by noticing how this is just iterating over [pp,pp1,pp2,pp3] but executing mechanical actions, so you can write something which runs a `forM`, calls internally `replicateM 2` , yields a `Property` and conjoin everything at the end.